File: scripts/get_edpub_token/get_token_cognito/python/get_token.py
#!/usr/bin/env python3
#===============================================================================
# AUTH CODE/IMPLICIT GRANTS, WITHOUT PKCE, WITHOUT CLIENT SECRET
#
# Source: https://stackoverflow.com/a/49015716
#===============================================================================

import requests
import json
import sys
import os
import re


## Set cognito variables
auth_domain = os.environ.get('AUTH_DOMAIN', 'edpub-prod.auth.us-west-2.amazoncognito.com')
client_id = os.environ.get('CLIENT_ID', 'aaev820uvfvon2227beng7d68')
response_type = os.environ.get('RESPONSE_TYPE', 'code')
redirect_uri = os.environ.get('REDIRECT_URI', 'https://pub.sit.earthdata.nasa.gov/dashboard/auth')
scope = os.environ.get('SCOPE', 'openid')
state = os.environ.get('STATE', 'dashboard')

cognito_user = os.environ.get('USERNAME')
cognito_pass = os.environ.get('PASSWORD')

# Check user specific variables
if not all([cognito_user, cognito_pass]):
    print('USERNAME and PASSWORD variables must be defined')
    sys.exit(1)

 # Create request session
session = requests.Session()

# Get xsrf token cookie
xsrf_response = session.get(f'https://{auth_domain}/oauth2/authorize', params={
    'response_type': response_type,
    'client_id' : client_id,
    'redirect_uri' : redirect_uri,
    'scope' : scope,
    'state' : state,
})


curl_redirect = xsrf_response.url
csrf_token = session.cookies.get_dict()['XSRF-TOKEN']

# Get auth code from /login endpoint
code_response = session.post(curl_redirect, headers={"cookie": f"XSRF-TOKEN={csrf_token}; Path=/; Secure; HttpOnly"}, data={
    "_csrf": f"{csrf_token}",
    "username": f"{cognito_user}",
    "password": f"{cognito_pass}"
})

auth_code = re.search(r'.*code=(.*)&.*', code_response.url)[1]

# Get tokens from /oauth2/token endpoint
grant_type = 'authorization_code'
token_dict = session.post(f'https://{auth_domain}/oauth2/token', data={
    'grant_type' : grant_type,
    'client_id' : client_id,
    'code' : auth_code,
    'redirect_uri' : redirect_uri
})
print(json.dumps(token_dict.json(), indent=4))

File: scripts/get_edpub_token/get_token_cognito/node/get_token.js
#!/usr/bin/node
// ===============================================================================
//  AUTH CODE/IMPLICIT GRANTS, WITHOUT PKCE, WITHOUT CLIENT SECRET
//
//  Source: https://stackoverflow.com/a/49015716
// ===============================================================================

import fetch from "node-fetch";

// Set cognito variables
const authDomain = process.env.AUTH_DOMAIN || 'edpub-prod.auth.us-west-2.amazoncognito.com';
const clientId = process.env.CLIENT_ID || 'aaev820uvfvon2227beng7d68';
const responseType = process.env.RESPONSE_TYPE || 'code';
const redirectURI = process.env.REDIRECT_URI || 'https://pub.sit.earthdata.nasa.gov/dashboard/auth';
const scope = process.env.SCOPE || 'openid';
const state = process.env.STATE || 'dashboard';

const cognitoUser = process.env.USERNAME || '';
const cognitoPass = process.env.PASSWORD || '';

// Check user specific variables
(!cognitoUser || !cognitoPass) && console.log('USERNAME and PASSWORD variables must be defined') && process.exit(1);

// Get xsrf token cookie
async function getXSRFResponse() {
    return await fetch(`https://${authDomain}/oauth2/authorize?response_type=${responseType}&client_id=${clientId}&redirect_uri=${redirectURI}&scope=${scope}&state=${state}`)
}

// Get auth code from /login endpoint
async function getAuthCode(url, token) {
    return await fetch(url, {
        method: 'POST',
        headers: {
            "content-type": "application/x-www-form-urlencoded",
            'cookie': `XSRF-TOKEN=${token}`,
        },
        body: `_csrf=${token}&username=${cognitoUser}&password=${encodeURIComponent(cognitoPass)}`
    }).then(res => res.url.match(/code=(.*)&state/)[1]);
}

// Get tokens from /oauth2/token endpoint
async function getAuthTokens(authCode) {
    const grantType = 'authorization_code';
    return await fetch(`https://${authDomain}/oauth2/token`, {
        body: `grant_type=${grantType}&client_id=${clientId}&code=${authCode}&redirect_uri=${redirectURI}`,
        headers: {
            "Content-Type": "application/x-www-form-urlencoded"
        },
        method: "POST"
    }).then(async res => await res.json())
}


async function getEDPubToken() {
    const xsrfResponse = await getXSRFResponse()
    const curlRedirect = xsrfResponse.url
    const csrfToken = xsrfResponse.headers.get('set-cookie').match(/XSRF-TOKEN=(.*); Path/)[1];
    const authCode = await getAuthCode(curlRedirect, csrfToken);
    console.log(await getAuthTokens(authCode))
}

getEDPubToken()
